### PR TITLE
Improve indexing runtime for serial data

### PR DESCRIPTION
**Pull request by jbeilstenedmands**
I noticed that indexing is by far the slowest step for processing serial data. After doing some profiling, I found that over half of the time was spent comparing against the known symmetry, more specifically, through many calls to the init method of the space_group_type object.

The existing code is doing:
```
For each crystal (~n images):
    For each model (~100-1000):
        calculate_target_bravais_t (same each time)
        calculate_subgroups (for_model_lattice_type)
        calculate_supergroups (for_model_lattice_type)
        calculate_cb_op (for_model_lattice_type)
        throw all this away if subgroup reference bravais_t doesn't match target_bravais_t
```
This PR moves `calculate_target_bravais_t` up one level. It also caches the subgroup, supergroup and cb_op calculation for a given lattice type, also skipping some further unnecessary calculations if the `bravais_t` does not match the target.

For indexing 1000s of images split over a number of processes (say 20), the caching avoids the need for many repeated `sgtbx` calculations and significantly improves runtime. The cache is separate for each process, which avoids any process locking when saving data to the cache. Results should be identical with the changes in this PR.

I timed `dials.stills_process` on 1000 images using 20 processes on our computing cluster, and found the overall runtime reduced from 299s -> 170s.
For my ssx processing script, the overall runtime using similar parameters is down from 362s -> 260s.
This was a cubic example, so the results might not be as impressive for lower symmetries, but there is a clear benefit!

It would be great if someone from @phyy-nx 's team was able to verify these kind of improvements on your computing setup, in case there is any funny behaviour with the multiprocessing and caching.
I would also appreciate a review from @rjgildea to see if he agrees that the new code is functionally equivalent. This function is called in all indexing tests that give a known symmetry, so this should already be well tested.

### CODE REVIEW

**Review by rjgildea — APPROVED** [review]
Thanks for looking at this - handling of symmetry in indexing is definitely ripe for performance improvements.
Did you look at using [`@functools.lru_cache`](https://docs.python.org/3/library/functools.html#functools.lru_cache) to implement the caching mechanism? This would probably simplify things somewhat if it works here.

**Comment by jbeilstenedmands** [comment]
Thanks for taking a look, I wasn't aware of `lru_cache` so I'll look into it now and test. I'm sure that there are more efficiency savings that could be made in indexing if we're interested in looking further, but this was by far the best in terms of the ratio of runtime improvement vs intrusiveness of changes.

**Comment by phyy-nx** [comment]
The xfel regression tests all pass.  I'd like to do a bit more testing with this later today.  Excited about better performance!

File: algorithms/indexing/symmetry.py [hunk 1]
```diff
@@ -23,13 +23,71 @@ def metric_supergroup(group):
...
+
+    instances = {}
+
+    def __new__(cls, target_bravais_t):
+        str_t = str(target_bravais_t)
+        if str_t not in cls.instances:
+            cls.instances[str_t] = _MatchingSubgroupsForLatticeGroup()
+        return cls.instances[str_t]
+
+
```
[the remark this answers is not in the document]
Comment (jbeilstenedmands) [reply algorithms/indexing/symmetry.py]: Thanks for the suggestion. I tried but couldn't use the `functools.lru_cache` easily as the arguments need to be hashable, and as this is a `sgtbx` object, you need to know how to convert to a suitable value (in this case with `.type().lookup_symbol()`). However, the idea was good, so I made a simple decorator to handle the caching logic, and I think this leads to much cleaner code in this case, so thanks for the idea!

**Comment by phyy-nx** [comment]
(haven't had a chance to test this on a big dataset yet due to beamtimes but I still want to.  hopefully soon, but I won't be bummed if you merged it sooner, since the current tests pass)

**Comment by jbeilstenedmands** [comment]
I'm happy to leave this open if you think you might get time to test it sometime this week, I'd like to make sure it's working for all multiprocessing setups before merging in.

**Comment by phyy-nx** [comment]
Hi, I was able to run my tests today.

I processed a recent modest LCLS run (10Hz) with a 23.5% indexing rate, netting 995 lattices (including multi-lattice search).  I tested it on main (plus some code from our recent beamtimes) and on this branch.  I used 4 nodes, with with 61 cores.  First, I verified identity in the results by using dials.combine_experiments on the two result folders, followed up by using this snippet:

```
from dials.array_family import flex
from dxtbx.model.experiment_list import ExperimentList
expts1 = ExperimentList.from_file('main/combined.expt', check_format=False)
expts2 = ExperimentList.from_file('branch/combined.expt', check_format=False)
list1 = sorted(expts1, key = lambda expt: expt.imageset.indices()[0])
list2 = sorted(expts2, key = lambda expt: expt.imageset.indices()[0])
for expt1, expt2 in zip(list1, list2):
    assert expt1.crystal == expt2.crystal
print('OK')
```

The sorting here lines up the results by image number and is needed due to how dials.stills_process uses client/server to split the data out into different MPI ranks.

Next I evaluated runtime using ```libtbx.python `libtbx.find_in_repositories xfel`/util/weather.py``` from the debug folders of the two results folders.  This indicated the overall runtime only sped up from 63 to 60s, but the median time per hit improved from 9.5 to 8.8s.  So on a dataset with a higher hitrate this would have a bigger impact.

Looks good to me :)
-Aaron

**Review by phyy-nx — APPROVED** [review]

**Comment by jbeilstenedmands** [comment]
Thanks @phyy-nx for confirming with your tests

**Pull request merged**